### ISSUE TITLE
Chef convergence fails on windows

### ISSUE THREAD

**Issue by mitom** [signalfx/signalfx-agent#598]
During the first convergence the cookbook fails on windows. The service gets created but it is stopped. It can be manually started and it will report data correctly. Probably a race condition?

```
Recipe: signalfx_agent::default
  * directory[\ProgramData\SignalFxAgent] action create[2019-02-25T14:06:04+00:00] INFO: Processing directory[\ProgramDa
ta\SignalFxAgent] action create (signalfx_agent::default line 23)
[2019-02-25T14:06:04+00:00] INFO: directory[\ProgramData\SignalFxAgent] created directory \ProgramData\SignalFxAgent

    - create new directory \ProgramData\SignalFxAgent[2019-02-25T14:06:04+00:00] INFO: directory[\ProgramData\SignalFxAg
ent] owner changed to S-1-5-21-2064249170-1864486422-3813938014-500
[2019-02-25T14:06:04+00:00] INFO: directory[\ProgramData\SignalFxAgent] group changed to S-1-5-21-2064249170-1864486422-
3813938014-500

    - change owner
    - change group
Recipe: signalfx_agent::win
  * windows_zipfile[\Program Files\SignalFx\] action unzip[2019-02-25T14:06:04+00:00] INFO: Processing windows_zipfile[\
Program Files\SignalFx\] action unzip (signalfx_agent::win line 2)

    - unzip https://dl.signalfx.com/windows/final/zip/SignalFxAgent-4.0.1-win64.zip
    * remote_file[C:/chef.cloudreach/cache/SignalFxAgent-4.0.1-win64.zip] action create[2019-02-25T14:06:04+00:00] INFO:
 Processing remote_file[C:/chef.cloudreach/cache/SignalFxAgent-4.0.1-win64.zip] action create (C:/chef.cloudreach/cache/
cookbooks/windows/resources/zipfile.rb line 40)
[2019-02-25T14:06:07+00:00] INFO: remote_file[C:/chef.cloudreach/cache/SignalFxAgent-4.0.1-win64.zip] created file C:/ch
ef.cloudreach/cache/SignalFxAgent-4.0.1-win64.zip

      - create new file C:/chef.cloudreach/cache/SignalFxAgent-4.0.1-win64.zip[2019-02-25T14:06:10+00:00] INFO: remote_f
ile[C:/chef.cloudreach/cache/SignalFxAgent-4.0.1-win64.zip] updated file contents C:/chef.cloudreach/cache/SignalFxAgent
-4.0.1-win64.zip

      - update content in file C:/chef.cloudreach/cache/SignalFxAgent-4.0.1-win64.zip from none to be6afe
      (file sizes exceed 10000000 bytes, diff output suppressed)
    * ruby_block[Unzipping] action run[2019-02-25T14:06:11+00:00] INFO: Processing ruby_block[Unzipping] action run (C:/
chef.cloudreach/cache/cookbooks/windows/resources/zipfile.rb line 54)
[2019-02-25T14:06:31+00:00] INFO: ruby_block[Unzipping] called

      - execute the ruby block Unzipping

  * file[\Program Files\SignalFx\\version.txt] action create[2019-02-25T14:06:31+00:00] INFO: Processing file[\Program F
iles\SignalFx\\version.txt] action create (signalfx_agent::win line 11)
[2019-02-25T14:06:31+00:00] INFO: file[\Program Files\SignalFx\\version.txt] created file \Program Files\SignalFx\\versi
on.txt

    - create new file \Program Files\SignalFx\\version.txt[2019-02-25T14:06:31+00:00] INFO: file[\Program Files\SignalFx
\\version.txt] updated file contents \Program Files\SignalFx\\version.txt

    - update content in file \Program Files\SignalFx\\version.txt from none to f4da6c
    --- \Program Files\SignalFx\\version.txt    2019-02-25 14:06:31.000000000 +0000
    +++ \Program Files\SignalFx/chef-version20190225-2604-x7pno1.txt    2019-02-25 14:06:31.000000000 +0000
    @@ -1 +1,2 @@
    +4.0.1
  * powershell_script[ensure service created] action run[2019-02-25T14:06:31+00:00] INFO: Processing powershell_script[e
nsure service created] action run (signalfx_agent::win line 23)
[2019-02-25T14:06:35+00:00] INFO: powershell_script[ensure service created] ran successfully

    - execute "C:\Windows\system32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NonInteractive -NoProfile -ExecutionP
olicy Bypass -InputFormat None -File "C:/Users/ADMINI~1/AppData/Local/Temp/2/chef-script20190225-2604-6g6g8d.ps1"
Recipe: signalfx_agent::default
  * template[\ProgramData\SignalFxAgent\agent.yaml] action create[2019-02-25T14:06:35+00:00] INFO: Processing template[\
ProgramData\SignalFxAgent\agent.yaml] action create (signalfx_agent::default line 49)
[2019-02-25T14:06:35+00:00] INFO: template[\ProgramData\SignalFxAgent\agent.yaml] created file \ProgramData\SignalFxAgen
t\agent.yaml

    - create new file \ProgramData\SignalFxAgent\agent.yaml[2019-02-25T14:06:35+00:00] INFO: template[\ProgramData\Signa
lFxAgent\agent.yaml] updated file contents \ProgramData\SignalFxAgent\agent.yaml

    - update content in file \ProgramData\SignalFxAgent\agent.yaml from none to 201e6b
    --- \ProgramData\SignalFxAgent\agent.yaml   2019-02-25 14:06:35.000000000 +0000
    +++ \ProgramData\SignalFxAgent/chef-agent20190225-2604-1iz4jrg.yaml 2019-02-25 14:06:35.000000000 +0000
    @@ -1 +1,26 @@
    +# Automatically generated by Chef
    +
    +---
    +hostname: xxxxx
    +ingestUrl: https://ingest.eu0.signalfx.com
    +apiUrl: https://api.eu0.signalfx.com
    +signalFxAccessToken: xxxxx
    +globalDimensions:
    +  chef_environment: mock-environment
    +  shortcode: TEST
    +monitors:
    +- type: cpu
    +- type: processlist
    +- type: filesystems
    +- type: disk-io
    +- type: memory
    +- type: net-io
    +- type: vmem
    +- type: host-metadata
    +observers:
    +- type: host
    +metricsToExclude:
    +- "#from": C:\Program Files\SignalFx\SignalFxAgent\lib\whitelist.json
    +  flatten: true
    +[2019-02-25T14:06:35+00:00] INFO: template[\ProgramData\SignalFxAgent\agent.yaml] owner changed to S-1-5-21-2064249
170-1864486422-3813938014-500
[2019-02-25T14:06:35+00:00] INFO: template[\ProgramData\SignalFxAgent\agent.yaml] group changed to S-1-5-21-2064249170-1
864486422-3813938014-500
[2019-02-25T14:06:35+00:00] INFO: template[\ProgramData\SignalFxAgent\agent.yaml] permissions changed to [WIN-H4KL0TTBQV
O\Administrator/flags:0/mask:c0010000]

    - change dacl
    - change owner
    - change group
[2019-02-25T14:06:35+00:00] INFO: template[\ProgramData\SignalFxAgent\agent.yaml] not queuing delayed action restart on
windows_service[signalfx-agent] (delayed), as it's already been queued
  * windows_service[signalfx-agent] action enable[2019-02-25T14:06:35+00:00] INFO: Processing windows_service[signalfx-a
gent] action enable (signalfx_agent::default line 57)


    ================================================================================
    Error executing action `enable` on resource 'windows_service[signalfx-agent]'
    ================================================================================

    SystemCallError
    ---------------
    The specified service does not exist as an installed service. - OpenService: The specified service does not exist as
 an installed service.

    Resource Declaration:
    ---------------------
    # In C:/chef.cloudreach/cache/cookbooks/signalfx_agent/recipes/default.rb

     57: service node['signalfx_agent']['service_name'] do
     58:   action [:enable, :start]
     59: end

    Compiled Resource:
    ------------------
    # Declared in C:/chef.cloudreach/cache/cookbooks/signalfx_agent/recipes/default.rb:57:in `from_file'

    windows_service("signalfx-agent") do
      action [:enable, :start]
      default_guard_interpreter :default
      service_name "signalfx-agent"
      enabled nil
      running nil
      masked nil
      pattern "signalfx-agent"
      declared_type :service
      cookbook_name "signalfx_agent"
      recipe_name "default"
    end

    System Info:
    ------------
    chef_version=14.10.9
    platform=windows
    platform_version=6.1.7601
    ruby=ruby 2.5.3p105 (2018-10-18 revision 65156) [x64-mingw32]
    program_name=C:/opscode/chef/bin/chef-client
    executable=C:/opscode/chef/bin/chef-client

[2019-02-25T14:06:35+00:00] INFO: Running queued delayed notifications before re-raising exception
[2019-02-25T14:06:35+00:00] INFO: template[C:\Program Files\SplunkUniversalForwarder/etc/system/local/inputs.conf] sendi
ng restart action to windows_service[splunk] (delayed)
Recipe: splunk::default
  * windows_service[splunk] action restart[2019-02-25T14:06:35+00:00] INFO: Processing windows_service[splunk] action re
start (splunk::default line 342)
[2019-02-25T14:06:37+00:00] INFO: windows_service[splunk] configured.
[2019-02-25T14:06:45+00:00] INFO: windows_service[splunk] restarted

    - restart service windows_service[splunk]
[2019-02-25T14:06:45+00:00] INFO: windows_zipfile[\Program Files\SignalFx\] sending restart action to windows_service[si
gnalfx-agent] (delayed)
Recipe: signalfx_agent::default
  * windows_service[signalfx-agent] action restart[2019-02-25T14:06:45+00:00] INFO: Processing windows_service[signalfx-
agent] action restart (signalfx_agent::default line 57)
[2019-02-25T14:06:45+00:00] INFO: windows_service[signalfx-agent] restarted

    - restart service windows_service[signalfx-agent]

Running handlers:
[2019-02-25T14:06:45+00:00] ERROR: Running exception handlers
  - CRChefReporting
Running handlers complete
[2019-02-25T14:06:45+00:00] ERROR: Exception handlers complete
Chef Client failed. 69 resources updated in 02 minutes 24 seconds
[2019-02-25T14:06:45+00:00] FATAL: Stacktrace dumped to C:/chef.cloudreach/cache/chef-stacktrace.out
[2019-02-25T14:06:45+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2019-02-25T14:06:45+00:00] FATAL: SystemCallError: windows_service[signalfx-agent] (signalfx_agent::default line 57) ha
d an error: SystemCallError: The specified service does not exist as an installed service. - OpenService: The specified
service does not exist as an installed service.
```

**Comment by keitwb** [comment]
I'm pretty sure this and #599 are both related to the service name being `SignalFx Smart Agent` in version 4.0.0 and 4.0.1.  It was changed to `signalfx-agent` to be consistent across platforms in 4.0.2.  You can either override the node['signalfx_agent']['service_name'] atttribute to be the old name or wait for 4.0.2 to be fully released.

**Comment by mitom** [comment]
resolved with 4.0.2 as explained above
